### ISSUE TITLE
Sorting a DataFrame and then creating a pivot_table for it produces wrong aggregate values

### ISSUE THREAD

**Issue by sebastianiorga** [SciRuby/daru#49]
``` ruby
$> df = Daru::DataFrame.new({
    a: [1,2,3,4,5,6]*100,  
    b: ['a','b','c','d','e','f']*100,  
    c: [11,22,33,44,55,66]*100,  
    d: ['r']*600  
})    
=> 
#<Daru::DataFrame:54383340 @name = d132569b-0e1e-4516-ae37-fc41c07bde58 @size = 600>
                    a          b          c          d 
         0          1          a         11          r 
         1          2          b         22          r 
         2          3          c         33          r 
         3          4          d         44          r 
         4          5          e         55          r 
         5          6          f         66          r 
         6          1          a         11          r 
         7          2          b         22          r 
         8          3          c         33          r 
         9          4          d         44          r 
        10          5          e         55          r 
        11          6          f         66          r 
        12          1          a         11          r 
        13          2          b         22          r 
        14          3          c         33          r 
       ...        ...        ...        ...        ... 

$> a = df.dup.sort([:b]).dup.pivot_table(index: [:b, :d], values: [:a, :c], agg: :sum)
=> 
#<Daru::DataFrame:56265240 @name = f7fd6275-b6a4-4ad9-b456-7f2aad0fc3ac @size = 6>
                    a          c 
["a", "r"]        346       3806 
["b", "r"]        350       3850 
["c", "r"]        354       3894 
["d", "r"]        346       3806 
["e", "r"]        350       3850 
["f", "r"]        354       3894 

$> b = df.dup.pivot_table(index: [:b, :d], values: [:a, :c], agg: :sum)
=> 
#<Daru::DataFrame:57262440 @name = 49eb6d38-ed2f-42f0-a764-d026d94eb18c @size = 6>
                    a          c 
["a", "r"]        100       1100 
["b", "r"]        200       2200 
["c", "r"]        300       3300 
["d", "r"]        400       4400 
["e", "r"]        500       5500 
["f", "r"]        600       6600 

```

I'm still chewing on what the cause is. I think the problem is related to the difference between how GroupBy#get_group and GroupBy#select_groups_from work. GroupBy#select_groups_from produces incorrect dataframes that look like what pivot_tale is making use of. #get_group produces correct results.

The code below is while inside a pry session triggered by a binding.pry inside GroupBy#apply_method.

``` ruby
$> self
=> #<Daru::Core::GroupBy:0x00000003d563c8
 @context=

#<Daru::DataFrame:32157200 @name = e15aa9fb-fbcf-45cd-ae01-44b8baf9fd0a @size = 600>
                    a          b          c          d 
       324          1          a         11          r 
       126          1          a         11          r 
       546          1          a         11          r 
       474          1          a         11          r 
       114          1          a         11          r 
       432          1          a         11          r 
       120          1          a         11          r 
       330          1          a         11          r 
       132          1          a         11          r 
       510          1          a         11          r 
       312          1          a         11          r 
       306          1          a         11          r 
       150          1          a         11          r 
       138          1          a         11          r 
       318          1          a         11          r 
       ...        ...        ...        ...        ... 
,
 @groups=
  {["a", "r"]=>
    [0,
     1,
     2,
     3,
     4,
     5,
     6,
     7,
     8,
     9,
     10,
     11,
     12,
     13,
     14,
$> first
=> 
#<Daru::DataFrame:26464220 @name = 9900ba68-a23c-454b-a821-4d04fab10eec @size = 6>
                    a          b          c          d 
         0          1          a         11          r 
       100          5          e         55          r 
       200          3          c         33          r 
       300          1          a         11          r 
       400          5          e         55          r 
       500          3          c         33          r 

$> get_group(['a', 'r'])
=> 
#<Daru::DataFrame:26321720 @name = f5fd645b-7769-4c75-a4b8-60c542d62a3e @size = 100>
                    a          b          c          d 
         0          1          a         11          r 
         1          1          a         11          r 
         2          1          a         11          r 
         3          1          a         11          r 
         4          1          a         11          r 
         5          1          a         11          r 
         6          1          a         11          r 
         7          1          a         11          r 
         8          1          a         11          r 
         9          1          a         11          r 
        10          1          a         11          r 
        11          1          a         11          r 
        12          1          a         11          r 
        13          1          a         11          r 
        14          1          a         11          r 
       ...        ...        ...        ...        ... 

```

Looking into it some more.

EDIT

I think I've pinpointed the issue to GroupBy#apply_method, see https://github.com/sebastianiorga/daru/commit/58b9dd15b027c902d2395957c42c0a3824bbdfd3. 

Basically the indices that GroupBy stores in the values of `@groups` are positional. For dataframes with the default numerical index of 0 -> numRows indexes actually behaves the same whether you access a position in the vector or the key in the vector(since the keys are 0 -> numRows as well). If you sort the dataframe then the index keys won't have values 0 -> numRows and things break down.

At least I think that's what's going on. The tests are still green on the fork and I'm now seeing correct behavior for pivot tables on sorted data frames, so I'll go with it.

**Comment by zverok** [comment]
On current daru:
```ruby
df = Daru::DataFrame.new({
           a: [1,2,3,4,5,6]*100,  
           b: ['a','b','c','d','e','f']*100,  
           c: [11,22,33,44,55,66]*100,  
           d: ['r']*600  
     })    
# => #<Daru::DataFrame(600x4)>
#       a   b   c   d
#   0   1   a  11   r
#   1   2   b  22   r
#   2   3   c  33   r
#   3   4   d  44   r
#   4   5   e  55   r
#   5   6   f  66   r
#   6   1   a  11   r
#   7   2   b  22   r
#   8   3   c  33   r
#   9   4   d  44   r
#  10   5   e  55   r
#  11   6   f  66   r
#  12   1   a  11   r
#  13   2   b  22   r
#  14   3   c  33   r
# ... ... ... ... ... 

a = df.dup.sort([:b]).dup.pivot_table(index: [:b, :d], values: [:a, :c], agg: :sum)
# => #<Daru::DataFrame(6x2)>
#              a    c
#    a    r  100 1100
#    b    r  200 2200
#    c    r  300 3300
#    d    r  400 4400
#    e    r  500 5500
#    f    r  600 6600 
```
